File: algopatterns/graphs/dfs_recursive.py
def dfs(node, adj, visited):
    visited[node] = True
    # -- INSERT ACTION ON NODE HERE --
    for neighbor in adj[node]:
        if not visited[neighbor]:
            dfs(neighbor, adj, visited)

if __name__ == "__main__":
    n = 0
    adj = [[] for _ in range(n)]
    visited = [False] * n
    dfs(0, adj, visited)

File: algopatterns/graphs/dfs_recursive.cpp
#include <iostream>
#include <vector>
using namespace std;

void dfs(int node, const vector<vector<int>>& adj, vector<bool>& visited) {
    visited[node] = true;
    // -- INSERT ACTION ON NODE HERE --
    for (int neighbor : adj[node])
        if (!visited[neighbor])
            dfs(neighbor, adj, visited);
}

int main() {
    int n = 0;
    vector<vector<int>> adj(n);
    vector<bool> visited(n, false);
    dfs(0, adj, visited);
    return 0;
}
